[ao] Added statistical threshold arg in Outlier Detector

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81369
* #81382
* #81252
* #81251
* #81250
* #81249
* __->__ #81174

Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Added statistical threshold arg in Outlier Detector

Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1071,7 +1090,9 @@ def _calculate_outlier_info(self, percentile_ratios: torch.Tensor, counted_batch
         # calculate whether channels were statistically significant
-        significant_size = [True if batch_size >= 30 else False for batch_size in num_batches_list]
+        significant_size = [
+            True if batch_size / total_batches >= self.statistical_threshold else False for batch_size in num_batches_list
```
Can simplify to `[batch_size / total_batches >= ... for batch_size in num_batches_list]`

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -938,6 +938,9 @@ class OutlierDetector(DetectorBase):
             Default: 0.975
+        statistical_threshold (float, optional): Fraction of batches to determine outliers per channel should be above this
+            Should be between 0 and 1
+            Default: 0.95
```
Hmm, should this be configurable? What is a potential use case?

andrewor14 replied:
Also this comment currently sounds like an exception will be thrown if the fraction is below this, but I don't think that's the case. Maybe add what happens if the fraction is below this, e.g. the result will be marked as "insignificant" or something

andrewor14 replied:
I see. To me that still sounds like a bit of a niche use case and I'm not sure if people would actually use it, but OK to keep this as is. One risk in making too many things configurable is an overly complicated API and extra backward compatibility burden.

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1051,18 +1066,22 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
         Returns a dictionary mapping:
             "outliers_detected" : list of bools per channel that are true if it is considered an outlier
-            "above_sample_threshold_count": if the per channel calculation had at least 30 samples (a random threshold)
+            "above_sample_threshold_count": if o_r was >= statistical_threshold:
```
The dict key name is now outdated I think

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -938,6 +938,11 @@ class OutlierDetector(DetectorBase):
         reference_percentile (float, optional): The denominator to find the relative scale of the 100th percentile
             Should be between 0 and 1
             Default: 0.975
+        statistical_threshold (float, optional): Fraction of batches to determine outliers per channel should be above this
```
Sorry I should've commented this last time but I feel the name can be more descriptive. Right now it sounds like it could be a threshold for anything, not necessarily related to fraction of batches. Any thoughts?

vspenubarthi replied:
changed to `fraction_batches_used_threshold` to be more descriptive

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1051,18 +1068,22 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
         Returns a dictionary mapping:
             "outliers_detected" : list of bools per channel that are true if it is considered an outlier
-            "above_sample_threshold_count": if the per channel calculation had at least 30 samples (a random threshold)
+            "sufficient_batches": if o_r was >= statistical_threshold:
```
This value is actually a fraction but `sufficient_batches` still makes it sound like the unit is number of batches. I feel you need "ratio" or "fraction" in there somewhere.

andrewor14 reviewed (approved):
Feel free to merge first and address the comments later if this is blocking. I just had a couple naming suggestions.

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added statistical threshold arg in Outlier Detector"


Summary: The outlier detector has a feature where it's able to notify
the user if below the whole set of batches that passed through were used
in Outlier calculation, which mainly happens as a result of 0-errors.
This changes the code so that instead of comparing against a value like
30 as we were before, we now let the user pass in an optional fractional
value and if the ratio of the batches used was below that value, the
detector alerts the user.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
